php/class-widget-posts.php

```diff
@@ -0,0 +1,770 @@
+	 *
+	 * @param string $id_base
+	 * @param array $new_instance
+	 * @param array $old_instance
+	 * @return array
+	 *
+	 * @throws Exception
+	 */
+	public function sanitize_instance( $id_base, $new_instance, $old_instance = array() ) {
+		if ( ! array_key_exists( $id_base, $this->widget_objs ) ) {
+			throw new Exception( "Unrecognized widget id_base: $id_base" );
+		}
+		if ( ! is_array( $new_instance ) ) {
+			throw new Exception( 'new_instance data must be an array' );
+		}
+		if ( ! is_array( $old_instance ) ) {
+			throw new Exception( 'old_instance data must be an array' );
+		}
+		$widget_obj = $this->widget_objs[ $id_base ];
+		// @codingStandardsIgnoreStart
+		$instance = @ $widget_obj->update( $new_instance, $old_instance ); // silencing errors because we can get undefined index notices
+		// @codingStandardsIgnoreEnd
+		return $instance;
+	}
+
+	/**
+	 * Create a new widget instance.
+	 *
+	 * @param string $id_base
+	 * @param array $instance
+	 * @return \WP_Post
+	 *
+	 * @throws Exception
+	 */
+	function insert_widget( $id_base, $instance = array() ) {
+		if ( ! array_key_exists( $id_base, $this->widget_objs ) ) {
+			throw new Exception( "Unrecognized widget id_base: $id_base" );
+		}
+
+		$instance = $this->sanitize_instance( $id_base, $instance );
+		$widget_number = $this->plugin->widget_number_incrementing->incr_widget_number( $id_base );
+		$post_arr = array(
+			'post_name' => "$id_base-$widget_number",
+			'post_content' => wp_json_encode( $instance, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES ), // For search indexing and post revision UI.
+			'post_content_filtered' => static::encode_post_content_filtered( $instance ),
+			'post_status' => 'publish',
+			'post_type' => static::INSTANCE_POST_TYPE,
+		);
+		$r = wp_insert_post( $post_arr, true );
+		if ( is_wp_error( $r ) ) {
+			throw new Exception( sprintf(
+				'Failed to insert/update widget instance "%s": %s',
+				$post_arr['post_name'],
+				$r->get_error_code()
+			) );
+		}
+		$post = get_post( $r );
+		return $post;
+	}
+
+	/**
+	 * Update an existing widget.
+	 *
+	 * @param string $widget_id
+	 * @param array $instance
+	 * @throws Exception
+	 * @return \WP_Post
+	 */
+	function update_widget( $widget_id, $instance = array() ) {
+		$parsed_widget_id = $this->plugin->parse_widget_id( $widget_id );
+		if ( empty( $parsed_widget_id ) ) {
+			throw new Exception( "Invalid widget_id: $widget_id" );
+		}
+		if ( ! $parsed_widget_id['widget_number'] || $parsed_widget_id['widget_number'] < 2 ) {
+			throw new Exception( "Widgets must start numbering at 2: $widget_id" );
+		}
+
+		$post_id = null;
+		$old_instance = array();
+		$post = $this->get_widget_post( $widget_id );
+		if ( $post ) {
+			$post_id = $post->ID;
+			try {
+				$old_instance = $this->get_widget_instance_data( $post );
+			} catch ( Exception $e ) {
+				$old_instance = array();
+			}
+		}
+		$instance = $this->sanitize_instance( $parsed_widget_id['id_base'], $instance, $old_instance );
```

Comment:
I finally discovered that `wp widget-posts migrate` is importing [incomplete widget instance data](https://www.diffchecker.com/ixqq2leb) for certain widgets because of `sanitize_instance()`. This is mostly due to how certain widgets use `$old_instance` as the primary set of data and update only certain attributes from the `$new_instance`. Considering that the instance data in the `option_widget_*` is already supposed to be safe, we can probably skip this additional sanitization until we can ensure that no data is lost?